core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/IrRowPatternToProgramRewriter.java

```diff
@@ -0,0 +1,278 @@
+    }
+
+    @Override
+    protected Integer visitIrLabel(IrLabel node, Void context)
+    {
+        instructions.add(new MatchLabel(labelMapping.get(node)));
+        size++;
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrEmptyPattern(IrEmptyPattern node, Void context)
+    {
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrAnchorPattern(IrAnchorPattern node, Void context)
+    {
+        switch (node.getType()) {
+            case PARTITION_START:
+                instructions.add(new MatchStart());
+                break;
+            case PARTITION_END:
+                instructions.add(new MatchEnd());
+                break;
+            default:
+                throw new IllegalStateException("unexpected anchor type: " + node.getType());
+        }
+        size++;
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrExcludedPattern(IrExcludedPattern node, Void context)
+    {
+        instructions.add(new Save());
+        size++;
+
+        size = process(node.getPattern());
+
+        instructions.add(new Save());
+        size++;
+
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrPatternAlternation(IrPatternAlternation node, Void context)
+    {
+        if (node.getPatterns().size() == 1) {
+            return process(node.getPatterns().get(0));
+        }
+
+        Split split = new Split(size + 1, -1);
+        instructions.add(split);
+        size++;
+
+        process(node.getPatterns().get(0));
+        Jump jump = new Jump(-1);
+        instructions.add(jump);
+        size++;
+        split.setSecond(size);
+
+        int end = process(new IrPatternAlternation(node.getPatterns().subList(1, node.getPatterns().size())));
+        jump.setTarget(end);
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrPatternConcatenation(IrPatternConcatenation node, Void context)
+    {
+        for (IrRowPattern pattern : node.getPatterns()) {
+            process(pattern);
+        }
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrPatternPermutation(IrPatternPermutation node, Void context)
+    {
+        Iterator<List<Integer>> permutations = orderedPermutations(IntStream.range(0, node.getPatterns().size())
+                .boxed()
+                .collect(toImmutableList()))
+                .iterator();
+
+        ImmutableList.Builder<IrRowPattern> concatenations = ImmutableList.builder();
+        while (permutations.hasNext()) {
+            List<Integer> permutation = permutations.next();
+            ImmutableList.Builder<IrRowPattern> parts = ImmutableList.builder();
+            for (int index : permutation) {
+                parts.add(node.getPatterns().get(index));
+            }
+            concatenations.add(new IrPatternConcatenation(parts.build()));
+        }
+        IrPatternAlternation alternation = new IrPatternAlternation(concatenations.build());
+
```

Comment:
Don't create artificial IR nodes here. If necessary, pull out the necessary logic to plan alternations and concatenations into a separate method and call them from here and from the other visitor methods.